File: solutions/problem_0027_remove_element/solution.cpp
/*
 * Problem 27: Remove Element
 * ==========================
 * Difficulty: Easy
 * Tags: Array, Two Pointers
 * Pattern: Two Pointers
 *
 * Time Complexity:  O(n)
 * Space Complexity: O(1)
 */

#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        // Two pointer approach - O(n) time, O(1) space
        int left = 0, right = nums.size() - 1;
        while (left < right) {
            int curr = nums[left] + nums[right];
            if (curr == val) {
                return {left, right};
            } else if (curr < val) {
                left++;
            } else {
                right--;
            }
        }
        return 0;
    }
};

File: solutions/problem_0027_remove_element/solution.py
"""
Problem 27: Remove Element
==========================
Difficulty: Easy
Tags: Array, Two Pointers
Pattern: Two Pointers

Time Complexity:  O(n)
Space Complexity: O(1)
"""

class Solution:
    def removeElement(self, nums: list[int], val: int) -> int:
        write = 0
        for num in nums:
            if num != val:
                nums[write] = num
                write += 1
        return write
